### PR TITLE
Weighted Directed Graph Validator Setup

### PR DESCRIPTION
**Pull request by TRY-ER**
## Description

Added a validator setup to validate proper formatting and value in string representation of polymer molecules

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/utils/poly_wd_graph_utils.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+from typing import List, Tuple
+import re
+
+
+class PolyWDGStringValidator():
+    """
+    Class for validating the string format of weighted directed graph data
```
Comment (rbharath) [inline deepchem/utils/poly_wd_graph_utils.py:7]: You need to document what this string format is with examples and references in the docs

File: docs/source/api_reference/utils.rst [hunk 2]
```diff
@@ -708,6 +708,22 @@ The utilities here are used to modify the attributes of the classes. Used by dif
...
+
+These classes and functions are required to handle converstion of string data to graph data 
+and validation of the same. 
+
+Polymer Weighted Directed Graph String Validator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This class validates the input string representation for graphical conversion of polymer data.
+It splits the input strings, validates their order and values. Upon recieving error it can return
+meaningful error message to the user indicating the type of error that hinders the conversion process.
```
Comment (rbharath) [inline docs/source/api_reference/utils.rst:722]: Add citations, examples. This is a new input format so it needs a lot of references/explanation

File: deepchem/utils/poly_wd_graph_utils.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
+from typing import List, Tuple
+import re
+
+
+class PolyWDGStringValidator():
+    """
+    Class for validating the string format of weighted directed graph data.
+    This class provides methods to validate the format of a datapoint string.
+    This is a specific string format that is used for storing weighted directed polymer data in a parsable format.
```
Comment (rbharath) [inline deepchem/utils/poly_wd_graph_utils.py:9]: Add a citation. Who invented this format?

File: docs/source/api_reference/utils.rst [hunk 4]
```diff
@@ -708,6 +708,31 @@ The utilities here are used to modify the attributes of the classes. Used by dif
...
 
+Polymer Weighted Directed Graph Data Utilities
+-----------------------------------------
+
+These classes and functions are required to handle converstion of string data to graph data 
+and validation of the same. 
+
+Polymer Weighted Directed Graph String Validator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
Comment (rbharath) [inline docs/source/api_reference/utils.rst:719]: Who invented this format? What is it? Is it standard?

File: deepchem/utils/poly_wd_graph_utils.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
```
Comment (GreatRSingh) [inline deepchem/utils/poly_wd_graph_utils.py:9]: pls move next line after 70 characters.

Comment (GreatRSingh) [reply deepchem/utils/poly_wd_graph_utils.py:9]: At few places (like code and the example format) its ok to go beyond 70 characters but mostly we prefer less than 70 per line in normal text.

File: deepchem/utils/poly_wd_graph_utils.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        This static method parses the datapoint string into 3 parts:
+        1. Monomer molecules
+        2. Fragments
+        3. Polymer rules
+
+        Parameters
+        ----------
+        datapoint : str
+            The datapoint string to parse
+        Returns
```
Comment (GreatRSingh) [inline deepchem/utils/poly_wd_graph_utils.py:43]: Pls leave a line between parameters and returns section.

File: deepchem/utils/poly_wd_graph_utils.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        List[str]
+            A list containing the parsed rule strings
+        """
+        if len(rules_str.split("<")) == 1:
+            raise ValueError(
+                "Invalid rules string: The rule string must contain '<' as a separator for rules !"
+            )
+        return rules_str.split("<")[1:]
+
+    def _validate_fragments(self, datapoint: str):
```
Comment (GreatRSingh) [inline deepchem/utils/poly_wd_graph_utils.py:79]: pls add doc string in this function.

File: deepchem/utils/poly_wd_graph_utils.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        monomer_mols, fragments, _ = self.get_parsed_vals(datapoint)
+        if len(fragments) != len(monomer_mols.split(".")):
+            raise ValueError(
+                f"Number of fragments and number of monomers should match. Mismatch -> No. of Fragments {len(fragments)} , No. of Monomers{len(monomer_mols.split('.'))}"
+            )
+
+    def _get_all_wildcards(self, text: str) -> List[str]:
```
Comment (GreatRSingh) [inline deepchem/utils/poly_wd_graph_utils.py:86]: same here, pls add doc string

File: deepchem/utils/poly_wd_graph_utils.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        3. The polymer rules should be in the correct format.
+        4. The atom indexes in the polymer rules should be valid and present in the monomer molecules string.
+
+        It raises ValueError if the string format is invalid.
+
+        Parameters
+        ----------
+        datapoint : str
+            The datapoint string to validate
+        Returns
```
Comment (GreatRSingh) [inline deepchem/utils/poly_wd_graph_utils.py:145]: pls leave an empty line between sections. it helps the deepchem docs render properly.

File: deepchem/utils/test/test_poly_wd_graph_utils.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+        polymer_rules = "<1-2:0.5:0.5"
+        polymer_rules = PolyWDGStringValidator.get_polymer_rules(polymer_rules)
+        assert polymer_rules == ["1-2:0.5:0.5"]
+
+    def test_valid_validate_function(self):
+        datapoint = "[1*]C.C[2*]|0.5|0.5|<1-2:0.5:0.5"
+        assert PolyWDGStringValidator().validate(datapoint)
+
+    def test_invalid_validate_function(self):
```
Comment (GreatRSingh) [inline deepchem/utils/test/test_poly_wd_graph_utils.py:35]: there is no assert function in this function can you pls add a comment explaining the reason, if it's necessary to not have assert function.

Comment (GreatRSingh) [reply deepchem/utils/test/test_poly_wd_graph_utils.py:35]: same comment for few more functions after this.

File: docs/source/api_reference/utils.rst [hunk 11]
```diff
@@ -708,6 +708,46 @@ The utilities here are used to modify the attributes of the classes. Used by dif
...
+The format is an extended string representation that is used as input format for Weighted Directed Message Passing Neural Network (wDMPNN) implementation
+using :code:`polymer-chemprop` python module. For more information and understanding, visit the `github repo <https://github.com/coleygroup/polymer-chemprop>`.
+The format development and implementation is done by Matteo Aldeghi and Connor W. Coley for their 
+work on "A graph representation of molecular ensembles for polymer property prediction". 
+
+The :code:`dc.utils.PolyWDGStringValidator` class is explicitly useful for validating a Weighted Directed
+Graph Representaion within a string data for polymers. It validates atom notations in monomer, valid Fragment
+weights for monomers, and valid polymer rules within the string representation.
+
+References:
```
Comment (GreatRSingh) [inline docs/source/api_reference/utils.rst:733]: we don't add references like this, we mostly add a one line reference.

**Review by GreatRSingh — CHANGES_REQUESTED** [review]
@TRY-ER I have requested a few changes pls go through them.

File: deepchem/utils/test/test_poly_wd_graph_utils.py (hunk 10 above)
Comment (TRY-ER) [reply deepchem/utils/test/test_poly_wd_graph_utils.py:35]: It checks for invalid format, so there is assertRaise. It tests that the given code returns expected errors when a certain kind of invalid format is provided.

Comment (TRY-ER) [reply deepchem/utils/test/test_poly_wd_graph_utils.py:35]: For each invalid input format there is no assert rather assert raises.

File: deepchem/utils/poly_wd_graph_utils.py (hunk 7 above)
Comment (TRY-ER) [reply deepchem/utils/poly_wd_graph_utils.py:79]: These are private methods, hence no doc strings were added. But if it's required, I will add it.

File: deepchem/utils/poly_wd_graph_utils.py (hunk 5 above)
Comment (TRY-ER) [reply deepchem/utils/poly_wd_graph_utils.py]: done

File: deepchem/utils/test/test_poly_wd_graph_utils.py (hunk 10 above)
Comment (TRY-ER) [reply deepchem/utils/test/test_poly_wd_graph_utils.py:35]:
```python
try:
    .....
except exception as e:
    .....
    assert e is ValueError
```
The above code block is equivalent to
```python
with self.assertRaise(valueError):
    ....
```
It just makes the code short and concise (inside the unittest)
Hence I have used it to validate the error raise.

File: deepchem/utils/poly_wd_graph_utils.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
+from typing import List, Tuple
+import re
+
+
+class PolyWDGStringValidator():
+    """
+    Class for validating the string format of weighted directed graph
```
Comment (rbharath) [inline deepchem/utils/poly_wd_graph_utils.py:7]: Add a citation for where this format comes from in the docstring

File: docs/source/api_reference/utils.rst [hunk 13]
```diff
@@ -708,6 +708,35 @@ The utilities here are used to modify the attributes of the classes. Used by dif
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This class validates the input string representation for graphical conversion of polymer data.
+It splits the input strings, validates their order and values. Upon recieving error it can return
+meaningful error message to the user indicating the type of error that hinders the conversion process.
+
+The format is an extended string representation that is used as input format for Weighted Directed Message Passing Neural Network (wDMPNN) implementation
+using :code:`polymer-chemprop` python module. For more information and understanding, visit the `github repo <https://github.com/coleygroup/polymer-chemprop>`.
+The format development and implementation is done by Matteo Aldeghi and Connor W. Coley for their 
+work on "A graph representation of molecular ensembles for polymer property prediction".
```
Comment (rbharath) [inline docs/source/api_reference/utils.rst:727]: Add a proper paper citation. Look up how to do this for rst files

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**